src/Http/Middleware/RateLimitMiddleware.php

```diff
@@ -0,0 +1,289 @@
+     * @return string
+     */
+    protected function getIdentifier(ServerRequestInterface $request): string
+    {
+        if ($this->config['identifierCallback'] !== null) {
+            return (string)call_user_func($this->config['identifierCallback'], $request);
+        }
+
+        return match ($this->config['identifier']) {
+            'ip' => $this->getClientIp($request),
+            'user' => $this->getUserIdentifier($request),
+            'route' => $this->getRouteIdentifier($request),
+            'api_key' => $this->getApiKeyIdentifier($request),
+            default => $this->getClientIp($request),
+        };
+    }
+
+    /**
+     * Get client IP address
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @return string
+     */
+    protected function getClientIp(ServerRequestInterface $request): string
+    {
+        $params = $request->getServerParams();
+
+        if (!empty($params['HTTP_CF_CONNECTING_IP'])) {
+            return $params['HTTP_CF_CONNECTING_IP'];
+        }
+
+        if (!empty($params['HTTP_X_FORWARDED_FOR'])) {
+            $ips = explode(',', $params['HTTP_X_FORWARDED_FOR']);
+
+            return trim($ips[0]);
+        }
+
+        if (!empty($params['HTTP_X_REAL_IP'])) {
+            return $params['HTTP_X_REAL_IP'];
+        }
+
+        return $params['REMOTE_ADDR'] ?? 'unknown';
+    }
+
+    /**
+     * Get user identifier
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @return string
+     */
+    protected function getUserIdentifier(ServerRequestInterface $request): string
+    {
+        $user = $request->getAttribute('identity');
+        if ($user) {
+            if (method_exists($user, 'getIdentifier')) {
+                return 'user_' . $user->getIdentifier();
+            } elseif (isset($user->id)) {
+                return 'user_' . $user->id;
+            }
+        }
+
+        return $this->getClientIp($request);
+    }
+
+    /**
+     * Get route identifier
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @return string
+     */
+    protected function getRouteIdentifier(ServerRequestInterface $request): string
+    {
+        $params = $request->getAttribute('params', []);
+        $route = sprintf(
+            '%s::%s.%s',
+            $params['plugin'] ?? 'app',
+            $params['controller'] ?? 'unknown',
+            $params['action'] ?? 'unknown',
+        );
+
+        return $route . '_' . $this->getClientIp($request);
+    }
+
+    /**
+     * Get API key identifier
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request
+     * @return string
+     */
+    protected function getApiKeyIdentifier(ServerRequestInterface $request): string
+    {
+        $apiKey = $request->getHeaderLine('X-API-Key');
+        if (empty($apiKey)) {
+            $apiKey = $request->getQuery('api_key', '');
```

Comment:
You'd need to use `getQueryParams()` here as `getQuery()` isn't part of the PSR standard.